File: 0417-pacific-atlantic-water-flow/0417-pacific-atlantic-water-flow.cpp
#include <vector>
#include <iostream>
using namespace std;

class Solution {
public:
    vector<vector<int>> pacificAtlantic(vector<vector<int>>& heights) {
        int m = heights.size();
        int n = heights[0].size();
        vector<vector<bool>> pacific(m, vector<bool>(n, false));
        vector<vector<bool>> atlantic(m, vector<bool>(n, false));
        vector<vector<int>> result;

        // 四個方向
        vector<pair<int, int>> dirs = {{1,0}, {-1,0}, {0,1}, {0,-1}};

        // 定義 DFS
        function<void(int,int,vector<vector<bool>>&)> dfs = [&](int r, int c, vector<vector<bool>>& visited) {
            visited[r][c] = true;
            for (auto [dr, dc] : dirs) {
                int nr = r + dr, nc = c + dc;
                if (nr < 0 || nc < 0 || nr >= m || nc >= n) continue;
                // 只往高度相等或更高的地方走（反向流入）
                if (!visited[nr][nc] && heights[nr][nc] >= heights[r][c]) {
                    dfs(nr, nc, visited);
                }
            }
        };

        // 太平洋邊界（上邊與左邊）
        for (int i = 0; i < m; i++) dfs(i, 0, pacific);
        for (int j = 0; j < n; j++) dfs(0, j, pacific);

        // 大西洋邊界（下邊與右邊）
        for (int i = 0; i < m; i++) dfs(i, n - 1, atlantic);
        for (int j = 0; j < n; j++) dfs(m - 1, j, atlantic);

        // 找交集
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (pacific[i][j] && atlantic[i][j]) {
                    result.push_back({i, j});
                }
            }
        }

        return result;
    }
};

File: 0417-pacific-atlantic-water-flow/0417-pacific-atlantic-water-flow.py
class Solution:
    def pacificAtlantic(self, heights):
        if not heights or not heights[0]:
            return []

        m, n = len(heights), len(heights[0])
        pacific_reachable = set()
        atlantic_reachable = set()

        directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]

        def dfs(r, c, visited, prev_height):
            # 超出邊界或已訪問過
            if (r < 0 or c < 0 or r >= m or c >= n or
                (r, c) in visited or heights[r][c] < prev_height):
                return
            visited.add((r, c))
            for dr, dc in directions:
                dfs(r + dr, c + dc, visited, heights[r][c])

        # 太平洋邊界（上邊、左邊）
        for i in range(m):
            dfs(i, 0, pacific_reachable, heights[i][0])
        for j in range(n):
            dfs(0, j, pacific_reachable, heights[0][j])

        # 大西洋邊界（下邊、右邊）
        for i in range(m):
            dfs(i, n - 1, atlantic_reachable, heights[i][n - 1])
        for j in range(n):
            dfs(m - 1, j, atlantic_reachable, heights[m - 1][j])

        # 取交集
        result = list(pacific_reachable & atlantic_reachable)
        return result
